src/rust/engine/rule_graph/src/lib.rs

```diff
@@ -786,31 +793,86 @@ pub fn params_str<T: TypeId>(params: &ParamTypes<T>) -> String {
-///
 pub fn entry_str<R: Rule>(entry: &Entry<R>) -> String {
-  match entry {
-    Entry::WithDeps(ref e) => entry_with_deps_str(e),
-    Entry::Param(type_id) => format!("Param({})", type_id),
+  entry_node_str_with_attrs(entry).entry_str
+}
+
+#[derive(Debug, Clone, PartialOrd, Ord, PartialEq, Eq)]
+pub struct GraphVizEntryWithAttrs {
+  entry_str: String,
+  attrs_str: Option<String>,
+}
+
+// This may not be the best abstraction.
+pub enum Palette {
+  Olive,
+  Gray,
+  Orange,
+  Blue,
+}
+
+impl Palette {
+  // https://c.eev.ee/kouyou/ is wonderful for selecting lovely color juxtapositions across multiple
+  // different color axes.
+  fn color_string(&self) -> String {
+    // These color values are all in HSV. See https://www.graphviz.org/doc/info/colors.html for
+    // other methods of specifying
+    // colors. https://renenyffenegger.ch/notes/tools/Graphviz/attributes/_color/index may also be
+    // useful.
+    match self {
+      Self::Olive => "0.2214,0.7179,0.8528".to_string(),
+      Self::Gray => "0.576,0,0.6242".to_string(),
+      Self::Orange => "0.08,0.5,0.976".to_string(),
+      Self::Blue => "0.5,1,0.9".to_string(),
+    }
+  }
+
+  pub fn color_format_string(&self) -> String {
+    format!("[color=\"{}\",style=filled]", self.color_string())
+  }
+}
+
+///
+/// Apply coloration to several nodes.
+pub fn entry_node_str_with_attrs<R: Rule>(entry: &Entry<R>) -> GraphVizEntryWithAttrs {
+  let (entry_str, attrs_str) = match entry {
+    &Entry::WithDeps(ref e) => (
+      entry_with_deps_str(e),
+      // Color "singleton" entries (with no params) as a nice olive green!
+      if e.params().is_empty() {
+        Some(Palette::Olive.color_format_string())
+      } else if let Some(color) = e.rule().and_then(|r| r.color()) {
+        // Color "intrinsic" entries (provided by the rust codebase) with a steely gray!
+        Some(color.color_format_string())
+      } else {
+        None
+      },
+    ),
+    &Entry::Param(type_id) => (
+      format!("Param({})", type_id),
+      // Color "Param"s with a laid-back orange!
+      Some(Palette::Orange.color_format_string()),
+    ),
+  };
+  GraphVizEntryWithAttrs {
+    entry_str,
+    attrs_str,
   }
 }
 
 fn entry_with_deps_str<R: Rule>(entry: &EntryWithDeps<R>) -> String {
   match entry {
-    EntryWithDeps::Inner(InnerEntry {
+    &EntryWithDeps::Inner(InnerEntry {
       ref rule,
       ref params,
-    }) => format!("{} for {}", rule, params_str(params)),
-    EntryWithDeps::Root(ref root) => {
+    }) => format!("{}\nfor {}", rule.fmt_for_graph(), params_str(params)),
+    &EntryWithDeps::Root(ref root) => format!(
       // TODO: Consider dropping this (final) public use of the keyword "Select", while ensuring
       // that error messages remain sufficiently grokkable.
-      format!(
-        "Select({}) for {}",
-        root.dependency_key,
-        params_str(&root.params),
-      )
-    }
+      "Select({})\nfor {}",
```

Comment:
I've never understood what `Select` means. Is it possible to remove this? If that's not easy to do, please punt on this because this PR is a huge improvement worth landing.

Reply:
No, it could be removed! @stuhood any objections to removing this last user-facing `Select()` string?